### PR TITLE
Avoid deserialization when accessing key/value [HZ-790]

### LINKED ISSUE #19427: SQL issues around deserializing into Object on the server [HZ-790]
While working on https://github.com/hazelcast/hazelcast/pull/19362 , I have noticed that we have still problems mostly around how to handle deserialization of Compact/Portable. 

The main thing is we need to deserialize into object (serializationService.toObject / GenericRecordQueryReader.read(path) ) only when it is needed. 

### Problem 1:

`GenericRecordQueryReader.read(path) ` is designed to be called on SQL queries where a nested field is accessed to be compared. 

Lets say that we have a Person and as its field Child where both of them are Portable/Compact. 
When a user does a select call as follows:
`SELECT * FROM map WHERE person.child > ? `
We need to call 
`GenericRecordQueryReader.read(path)` where it will return Child class if it is in the classpath. Otherwsie it will return GenericRecord and the comparison will fail, which is expected. 

We have places that we call `GenericRecordQueryReader.read(path)` wrongly. An example is the place where we extract field from value to return a response to the caller(e.g. client). In this path, only thing we need is genericRecord. Constructing the object is both an heavy operation and can cause correctness problems. The following test fails when there is a config on the cluster that does not match to the client(different version of the class exists on the cluster)

```java
 Config config = new Config();
        config.getJetConfig().setEnabled(true);
        CompactSerializationConfig compactSerializationConfig =
                config.getSerializationConfig().getCompactSerializationConfig();
        compactSerializationConfig.setEnabled(true);
        // registering this class to the member to see it does not affect any of the tests.
        // It has a different schema than all the tests
        compactSerializationConfig.register(Person.class, PERSON_TYPE_NAME, new CompactSerializer<Person>() {
            @Nonnull
            @Override
            public Person read(@Nonnull CompactReader in) {
                Person person = new Person();
                person.surname = in.readString("surname", "NotAssigned");
                return person;
            }

            @Override
            public void write(@Nonnull CompactWriter out, @Nonnull Person person) {
                out.writeString("surname", person.surname);
            }
        });

        ClientConfig clientConfig = new ClientConfig();
        clientConfig.getSerializationConfig().getCompactSerializationConfig().setEnabled(true);
        initializeWithClient(1, config, clientConfig);
        sqlService = client().getSql();
        
        String name = randomName();
        sqlService.execute("CREATE MAPPING " + name + " ("
                + "key_id INT "
                + ", nested OBJECT "
                + ") "
                + "TYPE " + IMapSqlConnector.TYPE_NAME + ' '
                + "OPTIONS ("
                + '\'' + OPTION_KEY_FORMAT + "'='" + COMPACT_FORMAT + '\''
                + ", '" + OPTION_KEY_COMPACT_TYPE_NAME + "'='" + PERSON_ID_TYPE_NAME + '\''
                + ", '" + OPTION_VALUE_FORMAT + "'='" + COMPACT_FORMAT + '\''
                + ", '" + OPTION_VALUE_COMPACT_TYPE_NAME + "'='" + PERSON_TYPE_NAME + '\''
                + ")"
        ).updateCount();

- 

        GenericRecord record = GenericRecordBuilder.compact(PERSON_TYPE_NAME)
                .setString("name", "John").build();
        client().getMap(name).put(1, GenericRecordBuilder.compact("parent")
                .setGenericRecord("nested", record).build());

        assertRowsAnyOrder(
                "SELECT * FROM " + name,
                singletonList(new Row(1, record))
        );
 ```

### Problem 2:
serializationService.toObject is used when we insert a nested object which causes following test to fail because again there is a different version of the class configured on the server. 

```java
        config.getJetConfig().setEnabled(true);
        CompactSerializationConfig compactSerializationConfig =
                config.getSerializationConfig().getCompactSerializationConfig();
        compactSerializationConfig.setEnabled(true);
        // registering this class to the member to see it does not affect any of the tests.
        // It has a different schema than all the tests
        compactSerializationConfig.register(Person.class, PERSON_TYPE_NAME, new CompactSerializer<Person>() {
            @Nonnull
            @Override
            public Person read(@Nonnull CompactReader in) {
                Person person = new Person();
                person.surname = in.readString("surname", "NotAssigned");
                return person;
            }

            @Override
            public void write(@Nonnull CompactWriter out, @Nonnull Person person) {
                out.writeString("surname", person.surname);
            }
        });
        
        String name = randomName();
        sqlService.execute("CREATE MAPPING " + name + " ("
                + "key_id INT EXTERNAL NAME \"__key.id\""
                + ", nested OBJECT "
                + ") "
                + "TYPE " + IMapSqlConnector.TYPE_NAME + ' '
                + "OPTIONS ("
                + '\'' + OPTION_KEY_FORMAT + "'='" + COMPACT_FORMAT + '\''
                + ", '" + OPTION_KEY_COMPACT_TYPE_NAME + "'='" + PERSON_ID_TYPE_NAME + '\''
                + ", '" + OPTION_VALUE_FORMAT + "'='" + COMPACT_FORMAT + '\''
                + ", '" + OPTION_VALUE_COMPACT_TYPE_NAME + "'='" + PERSON_TYPE_NAME + '\''
                + ")"
        ).updateCount();

        GenericRecord record = GenericRecordBuilder.compact(PERSON_TYPE_NAME)
                .setString("name", "John").build();
        sqlService.execute("SINK INTO " + name + " VALUES (1, ?)",
                record).updateCount();

        assertRowsAnyOrder(
                "SELECT * FROM " + name,
                singletonList(new Row(1, record))
        );
```

There are some existing tests on the codebase to verify some of the problems described here: 

- `SqlClientCompactQueryTest.testQueryOnObject`.  Currently ignored. 
- `SqlCompactTest`. In the setup part, a config is deliberatly left commented. 
```java
 // registering this class to the member to see it does not affect any of the tests.
        // It has a different schema than all the tests
        compactSerializationConfig.register(Person.class, PERSON_TYPE_NAME, new CompactSerializer<Person>() {
            @Nonnull
            @Override
            public Person read(@Nonnull CompactReader in) {
                Person person = new Person();
                person.surname = in.readString("surname", "NotAssigned");
                return person;
            }

            @Override
            public void write(@Nonnull CompactWriter out, @Nonnull Person person) {
                out.writeString("surname", person.surname);
            }
        });
```
If this code is added back some of the tests fail. Namely, 
`test_topLevelFieldExtraction`
`test_insertingNestedCompact`
`test_selectNestedCompact`

We need to add more tests around similar use-cases. Some problems could also exist on the Portable. It is needed to be tested as well.
Mainly, usages around `LazyTarget`, `CompactGetter`, `serializationService.toObject` is needed to be examined.

### PR DESCRIPTION
**Pull request by viliam-durina**
Fixes #19427

### CODE REVIEW

**Comment by viliam-durina** [comment]
run-lab-run

**Comment by viliam-durina** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/sql/impl/row/JetSqlRow.java [hunk 1]
```diff
@@ -1,30 +1,29 @@
 /*
- * Copyright 2021 Hazelcast Inc.
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
  *
- * Licensed under the Hazelcast Community License (the "License");
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (ivanthescientist) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/row/JetSqlRow.java:4]: Is this intentional? Aren't we using HZ license everywhere?

**Review by ivanthescientist — APPROVED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/JetSqlSerializerHook.java [hunk 2]
```diff
@@ -65,7 +63,6 @@ public DataSerializableFactory createFactory() {
         constructors[JSON_VALUE] = arg -> new JsonValueFunction<>();
         constructors[JSON_OBJECT] = arg -> new JsonObjectFunction();
         constructors[JSON_ARRAY] = arg -> new JsonArrayFunction();
-        constructors[JET_SQL_ROW] = arg -> new JetSqlRow();
```
Comment (sancar) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/JetSqlSerializerHook.java:68]: Should we guarantee 5.0 and 5.1 members working together for SQL ? If yes, this looks like a non backward compatible change.

Comment (viliam-durina) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/JetSqlSerializerHook.java:68]: They work together in a special way: one query is always executed only by a subset of members with equal cluster version. Therefore, `JetSqlRow` instances will always be sent only to members of the same version.

**Review by sancar — APPROVED** [review]

**Review by Fly-Style — APPROVED** [review]

**Pull request merged**